synapse/lib/lmdb.py

```diff
@@ -553,3 +565,411 @@ def encodeValAsKey(v, isprefix=False):
             return _STR_VAL_MARKER + v_enc
         else:
             return _STR_VAL_MARKER + v_enc + b'\x00'
+
+class Slab(s_base.Base):
+    '''
+    A "monolithic" LMDB instance for use in a asyncio loop thread.
+    '''
+    COMMIT_PERIOD = 1.0  # time between commits
+
+    async def __anit__(self, path, **opts):
+        await s_base.Base.__anit__(self)
+
+        self.path = path
+        self.optspath = os.path.join(path, 'opts.json')
+
+        if os.path.isfile(self.optspath):
+            opts.update(s_common.jsload(self.optspath))
+
+        self.mapsize = opts.get('map_size')
+        if self.mapsize is None:
+            raise Exception('Slab requires map_size!')
+
+        # save the transaction deltas in case of error...
+        self.xactops = []
+        self.recovering = False
+
+        # if growsize is not set, we double...
+        self.maxsize = opts.pop('maxsize', None)
+        self.growsize = opts.pop('growsize', None)
+
+        opts.setdefault('max_dbs', 128)
+        opts.setdefault('writemap', True)
+
+        self.lenv = lmdb.open(path, **opts)
+
+        self.scans = set()
+
+        self.holders = 0
+
+        self._initCoXact()
+
+        self.onfini(self._onCoFini)
+        self.schedCoro(self._runSyncLoop())
+
+    def _saveOptsFile(self):
+        opts = {'map_size': self.mapsize, 'growsize': self.growsize}
+        s_common.jssave(opts, self.optspath)
+
+    async def _runSyncLoop(self):
+        while not self.isfini:
+            await self.waitfini(timeout=self.COMMIT_PERIOD)
+            if self.isfini:
+                # There's no reason to forcecommit on fini, because there's a separate handler to already do that
+                break
+            if self.holders == 0:
+                self.forcecommit()
+
+    async def _onCoFini(self):
+        assert s_glob.plex.iAmLoop()
+        self._finiCoXact()
+        self.lenv.close()
+        del self.lenv
+
+    def _finiCoXact(self):
+        assert s_glob.plex.iAmLoop()
+
+        [scan.bump() for scan in self.scans]
+
+        self.xact.commit()
+
+        self.xactops.clear()
+
+        del self.xact
+
+    def grow(self, size=None):
+        '''
+        Close out the current transaction and resize the memory map.
+        '''
+        with self._noCoXact():
+            self._growMapSize(size=size)
+
+    def _growMapSize(self, size=None):
+
+        if size is not None:
+            self.mapsize += size
+
+        elif self.growsize is not None:
+            self.mapsize += self.growsize
+
+        else:
+            self.mapsize *= 2
```

Comment:
Do we know if a incremental growisize will work during a large btree rebalancing?  in my experience large lmdb instances start growing by powers of two in terms of size.